### PR TITLE
fix synthetic bridged methods in 18w50a

### PR DESCRIPTION
**Pull request by asiekierka**
Requesting a review from @Chocohead.

### CODE REVIEW

File: mappings/18w50a.tiny [hunk 1]
```diff
@@ -8446,19 +8446,19 @@ METHOD	azb	(Lazb;)Lfh;	b	method_8140
 METHOD	azb	(Lazb;)Lawo;	c	method_8138
 CLASS	azb$a	net/minecraft/class_1867$class_1868
 METHOD	azb$a	(Lcom/google/gson/JsonArray;)Lfh;	a	method_8144
-METHOD	azb$a	(Lit;Lazb;)V	a	method_8143
+METHOD	azb$a	(Lit;Lazb;)V	a	method_8124
 METHOD	azb$a	(Lqc;Lcom/google/gson/JsonObject;)Lazb;	a	method_8142
 METHOD	azb$a	(Lqc;Lit;)Lazb;	a	method_8141
```
Comment (Chocohead) [inline mappings/18w50a.tiny:8451]: Example of hidden bridges here, this should be `method_8122`

**Review by Chocohead — CHANGES_REQUESTED** [review]
This catches all the explicit bridges but doesn't cover all cases where methods are bridges. For example:
```java
public interface ITest<T extends Number> {
	void eat(T number);

	T make();
}

public class Test implements ITest<Short> {

	@Override
	public void eat(Short number) {
		System.out.println("Yum, " + number);
	}

	@Override
	public Short make() {
		return Short.MAX_VALUE;
	}
}
```
Compiling will result in both `eat` and `make` having explicit bridges (with both synthetic and bridge flags) which take/return `Number`. The both bridges have the same names as the methods they bridge to, naughty in Java world but fine for the JVM.
When Mojang obfuscates, `eat` will keep it's bridge flag (and its name too from every case I've seen) but `make` will not (it'll get a new name as well). Potentially this is a side effect from ProGuard having the setting to avoid obfuscating methods with different return types to the same name (obeying Java's restriction the JVM ignores). Either way, this results in methods which are only marked as synthetic but are actually bridges in disguise. These implicit bridges will have different Notch/official names to the methods they bridge to, but should share the same Intermediary mapping as they are in effect the same method.

They're an utter pain to immediately see, I only know to look for it from Voldemap's decompile recompiling (which adds all the missing bridge flags back) producing slightly different method flags when I was making a Notch -> Voldemap crash reverser. You could probably catch them from the signatures matching a generic superclass's (non-final) method and the actual method being a single call/invoke and return with bogus line numbers (typically the very top of the class).

**Comment by asiekierka** [comment]
I see. However, is this alright to merge? It still fixes *some* instances of the problem, and we already agreed that breaking early is better than breaking late...

**Comment by Chocohead** [comment]
I haven't checked all 646 fixes from lack of time today, but it looks fine and fixing some of the problem is definitely better than none of it.

**Comment by modmuss50** [comment]
I believe an actual example of this still happening is Vec3i.method_10265 a fix for this does not show in the diff in this PR

**Comment by Chocohead** [comment]
Yeah, that's a prime example of a hidden bridge. The Java interface doesn't get renamed but still loses the bridge flag whilst the concrete method gets renamed. Not sure why ProGuard wants to strip the bridge there when it can't do anything else, but apparently that's what it does. No doubt you'll see similar things from classes that override methods in `Collection`.

There are examples of both the bridge and the actual method being named as well, like [`StringTag`](https://github.com/FabricMC/yarn/blob/18w50a/mappings/net/minecraft/nbt/StringTag.mapping#L13-L14). If you extend `StringTag` you don't have to implement `Tag#copy` which is part of why the names being the same is helpful.

**Pull request closed without merging**